File: mac/LRActionKit/LRActionKit/Action/Manifest/LRManifestLayer.m
@import LRCommons;
@import PackageManagerKit;

#import "LRManifestLayer.h"
#import "LRManifestErrorSink.h"
#import "ActionKitSingleton.h"


@interface LRManifestLayer ()

@end


@implementation LRManifestLayer

- (instancetype)initWithManifest:(NSDictionary *)manifest errorSink:(id<LRManifestErrorSink>)errorSink {
    return [self initWithManifest:manifest requiredPackageReferences:[LRManifestLayer packageReferencesWithManifest:manifest] errorSink:errorSink];
}

- (instancetype)initWithManifest:(NSDictionary *)manifest requiredPackageReferences:(NSArray *)requiredPackageReferences errorSink:(id<LRManifestErrorSink>)errorSink {
    self = [super initWithManifest:manifest errorSink:errorSink];
    if (self) {
        _packageReferences = requiredPackageReferences;
    }
    return self;
}

+ (NSDictionary *)manifestByMergingLayers:(NSArray *)layers {
    NSMutableDictionary *result = [NSMutableDictionary new];
    for (LRManifestLayer *layer in layers) {
        [result setValuesForKeysWithDictionary:layer.manifest];
    }
    return [result copy];
}

+ (NSArray *)packageReferencesWithManifest:(NSDictionary *)manifest {
    LRPackageManager *packageManager = [ActionKitSingleton sharedActionKit].packageManager;
    return [manifest[@"applies_to"] arrayByMappingElementsUsingBlock:^id(NSDictionary *packageInfo) {
        return [packageManager packageReferenceWithDictionary:packageInfo];
    }];
}

@end

File: mac/LRActionKit/LRActionKit/Action/Manifest/LRManifestLayer.swift
import Foundation
import SwiftyFoundation
import LRCommons
import PackageManagerKit

public class LRManifestLayer: LRManifestBasedObject {

    // parsed by the layers to seed the package version detection system
    public let packageReferences: [LRPackageReference]

    public convenience override init(manifest: [String: AnyObject], errorSink: LRManifestErrorSink?) {
        let packageManager = ActionKitSingleton.sharedActionKit().packageManager;

        let dicts = ArrayValue(manifest["applies_to"], { $0 as? [String: AnyObject] }) ?? []
        let packageReferences = mapIf(dicts) { packageManager.packageReferenceWithDictionary($0) }

        self.init(manifest: manifest, requiredPackageReferences: packageReferences, errorSink: errorSink)
    }

    public init(manifest: [String: AnyObject], requiredPackageReferences: [LRPackageReference], errorSink: LRManifestErrorSink?) {
        self.packageReferences = requiredPackageReferences
        super.init(manifest: manifest, errorSink: errorSink)
    }

}
